### PR TITLE
Added combat status

### PR DESCRIPTION
**Pull request by kywch**
* `config.COMBAT_STATUS_DURATION` was added
* `Player` class (entity/player.py): `latest_combat_tick` and `in_combat()` were added
* `Attack` action updates both the attacker and target's latest_combat_tick, if the action is valid
* `Use`, `Destroy`, `Sell`, `Give`, `Buy`, `Give-Gold` check the agent's `in_combat`, and ignore the action if `in_combat` is True
* `ActionTargets` in `Observation` also checks the agent's `in_combat` and generate the full mask if in combat.
* added tests

### CODE REVIEW

File: nmmo/core/env.py [hunk 1]
```diff
@@ -366,7 +366,8 @@ def _compute_observations(self):
 
       obs[agent_id] = Observation(
         self.config, self.realm.tick,
-        agent_id, visible_tiles, visible_entities, inventory, market)
+        agent_id, visible_tiles, visible_entities, inventory, market,
+        agent.in_combat)
```
Comment (daveey) [inline nmmo/core/env.py]: i don't think we want to pass this directly to obs, shoudn't this just be a column on EntityState?

File: nmmo/io/action.py [hunk 2]
```diff
@@ -275,6 +275,11 @@ def call(realm, entity, style, target):
     if style.freeze and dmg > 0:
       target.status.freeze.update(config.COMBAT_FREEZE_TIME)
 
+    # record the combat tick for both player entities
+    for ent in [entity, target]:
+      if ent.is_player:
+        ent.latest_combat_tick = realm.tick + 1 # because the tick is about to increment
+
```
Comment (daveey) [inline nmmo/io/action.py:282]: this won't catch cases where an NPC attacks the player. we should be setting the tick in enity.receiveDamage and entity.giveDamage

**Review by daveey — CHANGES_REQUESTED** [review]

File: nmmo/io/action.py (hunk 2 above)
Comment (kywch) [reply nmmo/io/action.py:282]: Attacks by NPCs can be caught. 

image

File: nmmo/core/env.py (hunk 1 above)
Comment (kywch) [reply nmmo/core/env.py]: `latest_combat_tick` is added to EntityState, and `Observation` computes whether this agent is in combat status. Other places have been modified accordingly.

**Comment by kywch** [comment]
I encountered a RecursionError during map generation, which happened while placing fish on the map, so I removed the recursion from the function.

**Review by daveey — APPROVED** [review]

**Pull request merged**